### ISSUE TITLE
Unexpected worker state transition fetch->memory

### ISSUE THREAD

**Issue by crusaderky** [dask/distributed#5169]
CC @fjetter

A test just randomly failed in CI:
https://github.com/dask/distributed/pull/5166/checks?check_run_id=3239515373

```
Traceback (most recent call last):
  File "/Users/runner/work/distributed/distributed/distributed/utils.py", line 638, in log_errors
    yield
  File "/Users/runner/work/distributed/distributed/distributed/worker.py", line 2451, in gather_dep
    self.transition(ts, "memory", value=data[d])
  File "/Users/runner/work/distributed/distributed/distributed/worker.py", line 1723, in transition
    func = self._transitions[start, finish]
KeyError: ('fetch', 'memory')
```

**Comment by fjetter** [comment]
See https://github.com/dask/distributed/issues/5152 and https://github.com/dask/distributed/pull/5157

closing as dup

**Comment by fjetter** [comment]
test failure should be fixed with #5157

**Comment by fjetter** [comment]
I think this is the very first time we were able to connect one of the observed deadlocks to a flaky test. Maybe there is a chance for stable CI after all :)
